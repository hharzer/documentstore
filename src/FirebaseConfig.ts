import { enforceMandatory } from './helpers';
import uuid = require('uuid');
export enum FirebaseMode {
  MEMORY = 'MEMORY',
  EMULATOR = 'EMULATOR',
  PROD = 'PROD'
}
class FirebaseConfig {
  static of(env: NodeJS.ProcessEnv): FirebaseConfig {
    enforceMandatory('Firebase', env, [this.FIREBASE_MODE]);
    const mode = env[this.FIREBASE_MODE];
    switch (mode) {
      case FirebaseMode.PROD:
        enforceMandatory('Firebase', env, this.KEYS);
        return new FirebaseConfig(
          mode,
          env[this.FIREBASE_DB_URL],
          env[this.FIREBASE_PROJECT_ID],
          env[this.FIREBASE_CLIENT_EMAIL],
          env[this.FIREBASE_PRIVATE_KEY],
          env[this.FIREBASE_APP_NAME]
        );
      case FirebaseMode.EMULATOR: {
        const name = uuid.v4();
        const project = uuid.v4();

        return new FirebaseConfig(mode, name, project, name, name, name);
      }
      case FirebaseMode.MEMORY: {
        return new FirebaseConfig(mode, null, null, null, null, null);
      }
    }
  }

  static FIREBASE_MODE = 'FIREBASE_MODE';
  static FIREBASE_DB_URL = 'FIREBASE_DB_URL';
  static FIREBASE_PROJECT_ID = 'FIREBASE_PROJECT_ID';
  static FIREBASE_CLIENT_EMAIL = 'FIREBASE_CLIENT_EMAIL';
  static FIREBASE_PRIVATE_KEY = 'FIREBASE_PRIVATE_KEY';
  static FIREBASE_APP_NAME = 'FIREBASE_APP_NAME';

  static KEYS = [
    FirebaseConfig.FIREBASE_DB_URL,
    FirebaseConfig.FIREBASE_PROJECT_ID,
    FirebaseConfig.FIREBASE_CLIENT_EMAIL,
    FirebaseConfig.FIREBASE_PRIVATE_KEY
  ];
  constructor(
    readonly mode: string,
    readonly dbUrl: string,
    readonly projectId: string,
    readonly clientEmail: string,
    readonly privateKey: string,
    readonly appName: string
  ) {}
}

export default FirebaseConfig;
