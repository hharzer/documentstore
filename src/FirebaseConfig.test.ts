import FirebaseConfig, { FirebaseMode } from './FirebaseConfig';

const DB_URL = 'db_url';
const PROJECT_ID = 'project_id';
const CLIENT_EMAIL = 'client_email';
const PRIVATE_KEY = 'private_key';
const APP_NAME = 'app_name';
const MODE = 'PROD';
const env: NodeJS.ProcessEnv = {
  FIREBASE_DB_URL: DB_URL,
  FIREBASE_PROJECT_ID: PROJECT_ID,
  FIREBASE_CLIENT_EMAIL: CLIENT_EMAIL,
  FIREBASE_PRIVATE_KEY: PRIVATE_KEY,
  FIREBASE_APP_NAME: APP_NAME,
  FIREBASE_MODE: MODE
};
describe('FirebaseConfig', () => {
  it('should extract all config information from env', () => {
    const config = FirebaseConfig.of(env);
    expect(config.dbUrl).toBe(DB_URL);
    expect(config.projectId).toEqual(PROJECT_ID);
    expect(config.clientEmail).toEqual(CLIENT_EMAIL);
    expect(config.privateKey).toEqual(PRIVATE_KEY);
    expect(config.appName).toEqual(APP_NAME);
  });
  it('should build emulator config', () => {
    const config = FirebaseConfig.of({ FIREBASE_MODE: FirebaseMode.EMULATOR });
    expect(config.dbUrl).toBeDefined();
    expect(config.projectId).toBeDefined();
    expect(config.clientEmail).toBeDefined();
    expect(config.privateKey).toBeDefined();
    expect(config.appName).toBeDefined();
  });

  FirebaseConfig.KEYS.forEach(k => {
    it(`should fail on missing ${k}`, () => {
      const invalidEnv = { ...env };
      delete invalidEnv[k];
      expect(() =>
        FirebaseConfig.of(invalidEnv)
      ).toThrowErrorMatchingSnapshot();
    });
    it(`should fail on null ${k}`, () => {
      const invalidEnv = { ...env };
      invalidEnv[k] = null;
      expect(() =>
        FirebaseConfig.of(invalidEnv)
      ).toThrowErrorMatchingSnapshot();
    });
    it(`should fail on empty ${k}`, () => {
      const invalidEnv = { ...env };
      invalidEnv[k] = '';
      expect(() =>
        FirebaseConfig.of(invalidEnv)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

export { env };
