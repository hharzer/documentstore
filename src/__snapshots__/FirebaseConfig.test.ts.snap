// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseConfig should fail on empty FIREBASE_CLIENT_EMAIL 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_CLIENT_EMAIL\\"
]"
`;

exports[`FirebaseConfig should fail on empty FIREBASE_DB_URL 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_DB_URL\\"
]"
`;

exports[`FirebaseConfig should fail on empty FIREBASE_PRIVATE_KEY 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_PRIVATE_KEY\\"
]"
`;

exports[`FirebaseConfig should fail on empty FIREBASE_PROJECT_ID 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_PROJECT_ID\\"
]"
`;

exports[`FirebaseConfig should fail on missing FIREBASE_CLIENT_EMAIL 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_CLIENT_EMAIL\\"
]"
`;

exports[`FirebaseConfig should fail on missing FIREBASE_DB_URL 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_DB_URL\\"
]"
`;

exports[`FirebaseConfig should fail on missing FIREBASE_PRIVATE_KEY 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_PRIVATE_KEY\\"
]"
`;

exports[`FirebaseConfig should fail on missing FIREBASE_PROJECT_ID 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_PROJECT_ID\\"
]"
`;

exports[`FirebaseConfig should fail on null FIREBASE_CLIENT_EMAIL 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_CLIENT_EMAIL\\"
]"
`;

exports[`FirebaseConfig should fail on null FIREBASE_DB_URL 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_DB_URL\\"
]"
`;

exports[`FirebaseConfig should fail on null FIREBASE_PRIVATE_KEY 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_PRIVATE_KEY\\"
]"
`;

exports[`FirebaseConfig should fail on null FIREBASE_PROJECT_ID 1`] = `
"Invalid Firebase configuration
missing or invalid keys : [
  \\"FIREBASE_PROJECT_ID\\"
]"
`;
